CorsixTH/Lua/world.lua

```diff
@@ -1267,56 +1272,68 @@ function World:getReputationImpact(hospital)
-  if control and (control[0].Random or control[0].Mean) then
-    -- Support standard values for mean and variance
-    local mean = control[0].Mean or 180
-    local variance = control[0].Variance or 30
-    -- How many days until next emergency?
-    local days = math.round(math.n_random(mean, variance))
-    local emergency_date = self.game_date:plusDays(days)
-
-    -- Make it the same format as for "controlled" emergencies
-    self.next_emergency_month = emergency_date:monthOfGame()
-    self.next_emergency_day = emergency_date:dayOfMonth()
-  else
-    if not self.next_emergency_no then
-      self.next_emergency_no = 0
-    else
-      repeat
-        self.next_emergency_no = self.next_emergency_no + 1
-        -- Level three is missing [5].
-        if not control[self.next_emergency_no] and
-            control[self.next_emergency_no + 1] then
-          self.next_emergency_no = self.next_emergency_no + 1
-        end
-      until not control[self.next_emergency_no] or
-            control[self.next_emergency_no].EndMonth >= self.game_date:monthOfGame()
+  if control[0].Random or control[0].Mean then
+    self:scheduleRandomEmergency(control)
+    return
+  end
+  repeat
+    self.next_emergency_no = self.next_emergency_no + 1
+    local emer_num = self.next_emergency_no
+    -- Account for missing Level 3 emergency[5]
+    if not control[emer_num] and control[emer_num + 1] then
+      emer_num = emer_num + 1
+      self.next_emergency_no = emer_num
     end
-
-    local emergency = control[self.next_emergency_no]
-
+    local emergency = control[emer_num]
     -- No more emergencies?
-    if not emergency or emergency.EndMonth == 0 then
+    if not emergency then
       self.next_emergency_month = 0
-    else
-      -- Generate the next month and day the emergency should occur at.
-      -- Make sure it doesn't happen in the past.
-      local start = math.max(emergency.StartMonth, self.game_date:monthOfGame())
-      local next_month = math.random(start, emergency.EndMonth)
-      self.next_emergency_month = next_month
-      local day_start = 1
-      if start == emergency.EndMonth then
-        day_start = self.game_date:dayOfMonth()
-      end
-      local day_end = Date(1, next_month):lastDayOfMonth()
-      self.next_emergency_day = math.random(day_start, day_end)
+      self.next_emergency_date = nil
+      return
     end
+  until self:computeNextEmergencyDates(emergency)
+end
+
+--! If a level file specifies random emergencies we make the next one as defined by the mean/variance given
+--!param control (table) Contains emergency information from level file
+function World:scheduleRandomEmergency(control)
+  -- Support standard values for mean and variance
+  local mean = control[0].Mean or 180
+  local variance = control[0].Variance or 30
+  -- How many days until next emergency?
+  local days = math.round(math.n_random(mean, variance))
+  local emergency_date = self.game_date:plusDays(days)
```

Comment:
There is potential that #422 issue still exists within this logic where a level config overrides mean and variance with bad values and the next emergency is scheduled in the past.

Reply:
So days needs to be +ve? Can probably do that easy